channel/erc20_depositor.go

```diff
@@ -40,42 +45,138 @@ const ERC20DepositorTXGasLimit = 100000
 // Return value of ERC20Depositor.NumTx.
 const erc20DepositorNumTx = 2
 
+// Keep track of the increase allowance and deposit processes.
+var mu sync.Mutex
+var locks = make(map[string]*sync.Mutex)
+
+// DepositResult is created to keep track of the returned values.
+type DepositResult struct {
+	Transactions types.Transactions
+	Error        error
+}
+
+// Create key from account address and asset to only lock the process when hub deposits the same asset at the same time.
+func lockKey(account common.Address, asset common.Address) string {
+	return fmt.Sprintf("%s-%s", account.Hex(), asset.Hex())
+}
+
+// Retrieves Lock for specific key.
+func handleLock(lockKey string) *sync.Mutex {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if lock, exists := locks[lockKey]; exists {
+		return lock
+	}
+
+	lock := &sync.Mutex{}
+	locks[lockKey] = lock
+	return lock
+}
+
+// Locks the lock argument, runs the given function and then unlocks the lock argument.
+func lockAndUnlock(lock *sync.Mutex, fn func()) {
+	mu.Lock()
+	defer mu.Unlock()
+	lock.Lock()
+	defer lock.Unlock()
+	fn()
+}
+
 // NewERC20Depositor creates a new ERC20Depositor.
 func NewERC20Depositor(token common.Address) *ERC20Depositor {
 	return &ERC20Depositor{Token: token}
 }
 
-// Deposit deposits ERC20 tokens into the ERC20 AssetHolder specified at the
-// request's asset address.
+// Deposit approves the value to be swapped and calls DepositOnly.
+//
+//nolint:funlen
 func (d *ERC20Depositor) Deposit(ctx context.Context, req DepositReq) (types.Transactions, error) {
-	// Bind a `AssetHolderERC20` instance.
-	assetholder, err := assetholdererc20.NewAssetholdererc20(req.Asset.EthAddress(), req.CB)
-	if err != nil {
-		return nil, errors.Wrapf(err, "binding AssetHolderERC20 contract at: %x", req.Asset)
-	}
+	lockKey := lockKey(req.Account.Address, req.Asset.EthAddress())
+	lock := handleLock(lockKey)
+
 	// Bind an `ERC20` instance.
 	token, err := peruntoken.NewPeruntoken(d.Token, req.CB)
 	if err != nil {
 		return nil, errors.Wrapf(err, "binding ERC20 contract at: %x", d.Token)
 	}
-	// Increase the allowance.
-	opts, err := req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
-	if err != nil {
-		return nil, errors.WithMessagef(err, "creating transactor for asset: %x", req.Asset)
+	callOpts := bind.CallOpts{
+		Pending: false,
+		Context: ctx,
 	}
-	tx1, err := token.IncreaseAllowance(opts, req.Asset.EthAddress(), req.Balance)
+	// variables for the return value.
+	var depResult DepositResult
+	var approvalReceived bool
+	var tx1 *types.Transaction
+	var err1 error
+	lockAndUnlock(lock, func() {
+		allowance, err := token.Allowance(&callOpts, req.Account.Address, req.Asset.EthAddress())
+		if err != nil {
```

Comment:
If any of these ifs should be true, which means that an error has occurred, the function will still continue to execute